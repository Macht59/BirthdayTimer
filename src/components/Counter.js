import React, { Component } from 'react';
import {
    AppRegistry,
    Text,
    StyleSheet,
    View,
    Dimensions,
} from 'react-native';

export default class Counter extends Component {

    constructor(props) {
        super(props);
        this.updateRemainingTime = this.updateRemainingTime.bind(this);
        this.initializeRemainingTime = this.initializeRemainingTime.bind(this);
        this.state = {
            countdownDate: null,
            birthDate: null,
        };

        setInterval(this.updateRemainingTime, 1000);
    }

    componentDidMount() {
        if (this.props.birthDate) {
            this.initializeRemainingTime();
        }
    }

    componentDidUpdate() {
        if (this.props.birthDate !== this.state.birthDate) {
            this.initializeRemainingTime();
        }
    }

    updateRemainingTime() {
        if (!this.state.countdownDate) {
            return;
        }

        this.setState(prevState => {
            return {
                countdownDate: new Date(prevState.countdownDate.getTime() - 1000),
            };
        });
    }

    initializeRemainingTime() {
        if (!this.props.birthDate) {
            this.setState(() => {
                return {
                    countdownDate: null,
                    birthDate: null,
                }
            })
            return;
        }

        let diff = this.props.birthDate - new Date().setFullYear(this.props.birthDate.getFullYear());
        if (diff < 0) {
            diff += 365 * 24 * 60 * 60 * 1000;
        }

        this.setState(() => {
            const dateWithTimeShift = new Date(diff);
            const pureDate = new Date(dateWithTimeShift.getTime() + dateWithTimeShift.getTimezoneOffset() * 60 * 1000);

            return {
                countdownDate: pureDate,
                birthDate: this.props.birthDate,
            };
        });
    }

    render() {
        return (<View style={styles.circle}>
            <View style={{
                flexDirection: "column",
                justifyContent: "center",
                alignItems: "center",
                flex: 1,
            }}>
                {!this.state.countdownDate && <Text style={styles.smallCircleText}>Please, swipe right</Text>}
                {!this.state.countdownDate && <Text style={styles.smallCircleText}>to select your birthdate</Text>}
                {this.state.countdownDate && <Text style={styles.circleText}>{this.state.countdownDate.getMonth()} Month</Text>}
                {this.state.countdownDate && <Text style={styles.circleText}>{this.state.countdownDate.getDate() - 1} Days</Text>}
                {this.state.countdownDate && <Text style={styles.circleText}>{this.state.countdownDate.getHours()} Hours</Text>}
                {this.state.countdownDate && <Text style={styles.circleText}>{this.state.countdownDate.getMinutes()} Minutes</Text>}
                {this.state.countdownDate && <Text style={styles.circleText}>{this.state.countdownDate.getSeconds()} Seconds</Text>}
            </View>
        </View>);
    }
}

const styles = StyleSheet.create({
    circle: {
        backgroundColor: 'skyblue',
        borderRadius: (Dimensions.get("window").width - 30) / 2,
        width: Dimensions.get("window").width - 30,
        height: Dimensions.get("window").width - 30,
        alignSelf: "center",
    },
    circleText: {
        color: "white",
        fontSize: 40,
        alignSelf: "center",
    },
    smallCircleText: {
        color: "white",
        fontSize: 30,
        alignSelf: "center",
    }
});

AppRegistry.registerComponent('Counter', () => Counter);