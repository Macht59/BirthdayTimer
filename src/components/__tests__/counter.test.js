import React from 'react';
import Counter from '../counter.component';
import renderer from 'react-test-renderer';

test('Counter component renders correctly', () => {
    const birthDate = new Date(151515);
    const rendered = renderer.create(<Counter birthDate={birthDate} />).toJSON();
    expect(rendered).toMatchSnapshot();
});