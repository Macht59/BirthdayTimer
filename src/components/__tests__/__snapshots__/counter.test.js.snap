// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "skyblue",
          "fontSize": 30,
          "textAlign": "center",
        },
        Object {
          "textShadowColor": "rgba(255, 255, 255, 2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "textShadowRadius": 20,
        },
        Object {
          "paddingBottom": 25,
          "paddingTop": 25,
          "textAlign": "center",
        },
      ]
    }
  >
    You will be
     
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      0
    </Text>
     
    years old in
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "skyblue",
          "borderRadius": 350,
          "height": 700,
          "justifyContent": "center",
          "width": 700,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 30,
            "textAlign": "center",
          }
        }
      >
        Please, swipe right to select your birthdate
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 15,
      }
    }
  >
    <RCTSwitch
      disabled={false}
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onValueChange={[Function]}
      style={
        Array [
          Object {
            "height": 31,
            "width": 51,
          },
          undefined,
        ]
      }
      value={false}
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "skyblue",
          "fontSize": 15,
          "paddingLeft": 15,
        }
      }
    >
      do not dim
    </Text>
  </View>
</View>
`;
